citro3d/src/render.rs

```diff
@@ -27,13 +80,194 @@ pub struct Target<'screen> {
+    #[doc(alias = "C3D_SetBufInfo")]
+    pub fn set_buffer_info(&mut self, buffer_info: &buffer::Info) {
+        let raw: *const _ = &buffer_info.0;
+        // LIFETIME SAFETY: C3D_SetBufInfo actually copies the pointee instead of mutating it.
+        unsafe { citro3d_sys::C3D_SetBufInfo(raw.cast_mut()) };
+    }
+
+    /// Get the attribute info being used, if it exists. Note that the resulting
+    /// [`attrib::Info`] is copied from the one currently in use.
+    #[doc(alias = "C3D_GetAttrInfo")]
+    pub fn attr_info(&self) -> Option<attrib::Info> {
+        let raw = unsafe { citro3d_sys::C3D_GetAttrInfo() };
+        attrib::Info::copy_from(raw)
+    }
+
+    /// Set the attribute info to use for any following draw calls.
+    #[doc(alias = "C3D_SetAttrInfo")]
+    pub fn set_attr_info(&mut self, attr_info: &attrib::Info) {
+        let raw: *const _ = &attr_info.0;
+        // LIFETIME SAFETY: C3D_SetAttrInfo actually copies the pointee instead of mutating it.
+        unsafe { citro3d_sys::C3D_SetAttrInfo(raw.cast_mut()) };
+    }
+
+    /// Render primitives from the current vertex array buffer.
+    #[doc(alias = "C3D_DrawArrays")]
+    pub fn draw_arrays(&mut self, primitive: buffer::Primitive, vbo_data: buffer::Slice<'pass>) {
+        self.set_buffer_info(vbo_data.info());
+
+        // TODO: should we also require the attrib info directly here?
         unsafe {
-            C3D_RenderTargetDelete(self.raw);
+            citro3d_sys::C3D_DrawArrays(
+                primitive as ctru_sys::GPU_Primitive_t,
+                vbo_data.index(),
+                vbo_data.len(),
+            );
+        }
+    }
+
+    /// Indexed drawing. Draws the vertices in `buf` indexed by `indices`.
+    #[doc(alias = "C3D_DrawElements")]
+    pub fn draw_elements<I: Index>(
+        &mut self,
+        primitive: buffer::Primitive,
+        vbo_data: buffer::Slice<'pass>,
+        indices: &Indices<'pass, I>,
+    ) {
+        self.set_buffer_info(vbo_data.info());
+
+        let indices = &indices.buffer;
+        let elements = indices.as_ptr().cast();
+
+        unsafe {
+            citro3d_sys::C3D_DrawElements(
+                primitive as ctru_sys::GPU_Primitive_t,
+                indices.len().try_into().unwrap(),
+                // flag bit for short or byte
+                I::TYPE,
+                elements,
+            );
         }
     }
+
+    /// Use the given [`shader::Program`] for subsequent draw calls.
+    pub fn bind_program(&mut self, program: &'pass shader::Program) {
+        // SAFETY: AFAICT C3D_BindProgram just copies pointers from the given program,
+        // instead of mutating the pointee in any way that would cause UB
+        unsafe {
+            citro3d_sys::C3D_BindProgram(program.as_raw().cast_mut());
+        }
+    }
+
+    /// Binds a [`LightEnv`] for the following draw calls.
+    pub fn bind_light_env(&mut self, env: Option<&'pass mut Pin<Box<LightEnv>>>) {
```

Comment:
Does this have to be a pin box?

Reply:
In fact, no, it's a remnant from the lighting implementation in which the `Instance` held the `LightEnv`. Thanks for catching this.